File: step_05_strings/5.2_medium/P001_sort_characters_by_frequency/solution.py
"""
Problem: Sort Characters by Frequency (LeetCode #451)
Difficulty: MEDIUM | XP: 25
"""
from typing import List, Optional
from collections import Counter


# ============================================================
# APPROACH 1: BRUTE FORCE -- HashMap + Sort
# Time: O(n log n)  |  Space: O(n)
#
# Count frequencies, sort unique chars by freq descending,
# rebuild string by repeating each char.
# ============================================================
def brute_force(s: str) -> str:
    freq = {}
    for c in s:
        freq[c] = freq.get(c, 0) + 1

    # Sort characters by frequency descending
    sorted_chars = sorted(freq.keys(), key=lambda c: -freq[c])

    result = []
    for c in sorted_chars:
        result.append(c * freq[c])

    return ''.join(result)


# ============================================================
# APPROACH 2: OPTIMAL -- HashMap + Bucket Sort
# Time: O(n)  |  Space: O(n)
#
# Use bucket sort: bucket[i] holds chars with frequency i.
# Iterate buckets from n down to 1 to build result.
# ============================================================
def optimal(s: str) -> str:
    freq = {}
    for c in s:
        freq[c] = freq.get(c, 0) + 1

    # Bucket sort: index = frequency
    n = len(s)
    buckets = [[] for _ in range(n + 1)]
    for c, count in freq.items():
        buckets[count].append(c)

    result = []
    for i in range(n, 0, -1):
        for c in buckets[i]:
            result.append(c * i)

    return ''.join(result)


# ============================================================
# APPROACH 3: BEST -- Counter + Bucket Sort (compact)
# Time: O(n)  |  Space: O(n)
#
# Same bucket sort logic, using Counter for conciseness.
# ============================================================
def best(s: str) -> str:
    freq = Counter(s)
    n = len(s)
    buckets = [[] for _ in range(n + 1)]
    for c, count in freq.items():
        buckets[count].append(c)
    return ''.join(c * i for i in range(n, 0, -1) for c in buckets[i])


if __name__ == "__main__":
    print("=== Sort Characters by Frequency ===\n")

    tests = [
        ("tree", "eert"),
        ("cccaaa", "aaaccc"),
        ("Aabb", "bbAa"),
        ("a", "a"),
    ]

    for s, expected in tests:
        print(f"Input:    \"{s}\"")
        print(f"Expected: \"{expected}\" (or valid variant)")
        print(f"Brute:    \"{brute_force(s)}\"")
        print(f"Optimal:  \"{optimal(s)}\"")
        print(f"Best:     \"{best(s)}\"")
        print()

File: step_05_strings/5.2_medium/P001_sort_characters_by_frequency/Solution.java
/**
 * Problem: Sort Characters by Frequency (LeetCode #451)
 * Difficulty: MEDIUM | XP: 25
 *
 * @author DSA_Nova
 */
import java.util.*;

public class Solution {

    // ============================================================
    // APPROACH 1: BRUTE FORCE -- HashMap + Sort
    // Time: O(n log n)  |  Space: O(n)
    //
    // Count frequencies, sort unique chars by freq descending,
    // rebuild string by repeating each char.
    // ============================================================
    static class BruteForce {
        public String frequencySort(String s) {
            Map<Character, Integer> freq = new HashMap<>();
            for (char c : s.toCharArray()) {
                freq.put(c, freq.getOrDefault(c, 0) + 1);
            }

            List<Character> chars = new ArrayList<>(freq.keySet());
            chars.sort((a, b) -> freq.get(b) - freq.get(a));

            StringBuilder sb = new StringBuilder();
            for (char c : chars) {
                int count = freq.get(c);
                for (int i = 0; i < count; i++) {
                    sb.append(c);
                }
            }
            return sb.toString();
        }
    }

    // ============================================================
    // APPROACH 2: OPTIMAL -- HashMap + Bucket Sort
    // Time: O(n)  |  Space: O(n)
    //
    // Use bucket sort: bucket[i] holds chars with frequency i.
    // Iterate buckets from n down to 1 to build result.
    // ============================================================
    static class Optimal {
        public String frequencySort(String s) {
            Map<Character, Integer> freq = new HashMap<>();
            for (char c : s.toCharArray()) {
                freq.put(c, freq.getOrDefault(c, 0) + 1);
            }

            int n = s.length();
            @SuppressWarnings("unchecked")
            List<Character>[] buckets = new List[n + 1];
            for (int i = 0; i <= n; i++) {
                buckets[i] = new ArrayList<>();
            }

            for (Map.Entry<Character, Integer> entry : freq.entrySet()) {
                buckets[entry.getValue()].add(entry.getKey());
            }

            StringBuilder sb = new StringBuilder();
            for (int i = n; i >= 1; i--) {
                for (char c : buckets[i]) {
                    for (int j = 0; j < i; j++) {
                        sb.append(c);
                    }
                }
            }
            return sb.toString();
        }
    }

    // ============================================================
    // APPROACH 3: BEST -- int[128] array + Bucket Sort
    // Time: O(n)  |  Space: O(n)
    //
    // Use int array instead of HashMap for faster counting.
    // Same bucket sort, but avoids HashMap overhead.
    // ============================================================
    static class Best {
        public String frequencySort(String s) {
            int[] freq = new int[128];
            for (char c : s.toCharArray()) freq[c]++;

            int n = s.length();
            @SuppressWarnings("unchecked")
            List<Character>[] buckets = new List[n + 1];
            for (int i = 0; i <= n; i++) buckets[i] = new ArrayList<>();

            for (int c = 0; c < 128; c++) {
                if (freq[c] > 0) buckets[freq[c]].add((char) c);
            }

            StringBuilder sb = new StringBuilder();
            for (int i = n; i >= 1; i--) {
                for (char c : buckets[i]) {
                    for (int j = 0; j < i; j++) sb.append(c);
                }
            }
            return sb.toString();
        }
    }

    public static void main(String[] args) {
        System.out.println("=== Sort Characters by Frequency ===\n");

        String[] tests = {"tree", "cccaaa", "Aabb", "a"};
        String[] expected = {"eert", "aaaccc", "bbAa", "a"};

        for (int t = 0; t < tests.length; t++) {
            String s = tests[t];
            System.out.println("Input:    \"" + s + "\"");
            System.out.println("Expected: \"" + expected[t] + "\" (or valid variant)");
            System.out.println("Brute:    \"" + new BruteForce().frequencySort(s) + "\"");
            System.out.println("Optimal:  \"" + new Optimal().frequencySort(s) + "\"");
            System.out.println("Best:     \"" + new Best().frequencySort(s) + "\"");
            System.out.println();
        }
    }
}
